### PR TITLE
Add LeanCode.Contracts.Admin

### PR DESCRIPTION
**Pull request by Komoszek**
`LeanCode.Contracts.Admin` is the backend part of the `Admin App Generator` (name WIP). This code is currently part of `contractsgenerator-typescript` repository, but maintaining it there might not be the best idea.

### CODE REVIEW

File: .github/workflows/contracts-admin.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      - name: Setup .NET SDK
+        uses: actions/setup-dotnet@v3
+        with:
+          dotnet-version: 7.0.x
+      - name: Restore
+        working-directory: src/LeanCode.Contracts.Admin
+        run: dotnet restore
+      - name: Generate Protobuf
+        working-directory: src/LeanCode.Contracts.Admin
+        run: SERVER_VERSION=2.0.0-alpha.4 ../../tools/generate.sh project --project=LeanCode.Contracts.Admin.csproj -o LeanCode.Contracts.Admin.pb
```
Comment (Saancreed) [inline .github/workflows/contracts-admin.yml:38]: I think it should be
```suggestion
        run: SERVER_VERSION=2.0.0-alpha.4 ../../tools/generate.sh project --project=LeanCode.Contracts.Admin.csproj -o LeanCode.Contracts.pb
```

File: src/LeanCode.Contracts.Admin/AdminQuery.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    public int PageSize { get; set; }
+
+    public SortOrderDTO? SortOrder { get; set; }
+    public string? SortBy { get; set; }
+}
+
+public enum SortOrderDTO
+{
+    Descending = 0,
+    Ascending = 1
```
Comment (Saancreed) [inline src/LeanCode.Contracts.Admin/AdminQuery.cs:15]: Can we invert this so the default is ascending? Seems more natural to me.

(But tbh I'd strongly prefer to instead have `bool Descending` property in the `AdminQuery` instead.)

File: src/LeanCode.Contracts.Admin/AdminQuery.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+public class AdminQueryResult<TResult>
+{
+    public long Total { get; set; }
+    public List<TResult> Items { get; set; }
+}
+
+public class AdminFilterRange<T>
+{
+    public T? From { get; set; }
+    public T? To { get; set; }
```
Comment (Saancreed) [inline src/LeanCode.Contracts.Admin/AdminQuery.cs:27]: Are those inclusive or exclusive? Might be a good idea to state this explicitly, either via better property names or just `<remarks>`.

File: src/LeanCode.Contracts.Admin/AdminQuery.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+[AttributeUsage(AttributeTargets.Property | AttributeTargets.Field, AllowMultiple = false)]
+public class AdminFilterFor : Attribute
+{
+    public AdminFilterFor(string name) { }
```
Comment (Saancreed) [inline src/LeanCode.Contracts.Admin/AdminQuery.cs:33]: Won't the value of `name` be lost after construction? You should probably preserve it in a field/property.

Same in attributes below.

**Review by Saancreed — COMMENTED** [review]
I hope it's not too late for feedback on contracts themselves :see_no_evil:

File: src/LeanCode.Contracts.Admin/AdminQuery.cs (hunk 3 above)
Comment (Komoszek) [reply src/LeanCode.Contracts.Admin/AdminQuery.cs:27]: It's not specified. Values of `From` and `To` are values that user inputs and how they are interpreted depends on the person writing query handler.

File: src/LeanCode.Contracts.Admin/AdminQuery.cs (hunk 2 above)
Comment (tomaszlaskowskiLC) [reply src/LeanCode.Contracts.Admin/AdminQuery.cs:15]: I agree, but maybe "**Sort**Descending"?

File: .github/workflows/contracts-admin.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        run: |
+          VERSION=${GITHUB_REF#refs/tags/contracts-admin-v}
+          if ! [[ ${VERSION} =~ ^[0-9]+\.[0-9]+\.[0-9]+(-[a-zA-Z0-9\.]+)?$ ]]
+          then
+            echo 1>&2 "Invalid version: ${VERSION}"
+            exit 1
+          fi
+          echo "version=${VERSION}" >> $GITHUB_OUTPUT
+      - name: Checkout
+        uses: actions/checkout@v2
```
Comment (Dragemil) [inline .github/workflows/contracts-admin.yml:26]: There's a `v4` iirc.

File: src/LeanCode.Contracts.Admin/AdminQuery.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
+﻿namespace LeanCode.Contracts.Admin;
+
+public abstract class AdminQuery<TResult> : IQuery<AdminQueryResult<TResult>>
+{
+    public int Page { get; set; }
```
Comment (Dragemil) [inline src/LeanCode.Contracts.Admin/AdminQuery.cs:5]: A remark whether it is zero or one based?

**Review by Saancreed — APPROVED** [review]

**Review by lukaszgarstecki — APPROVED** [review]

**Review by tomaszlaskowskiLC — APPROVED** [review]

**Review by Dragemil — APPROVED** [review]

**Pull request merged**